apkenv.c

```diff
@@ -88,20 +88,36 @@ read_file_impl(const char *filename, char **buffer, size_t *size)
 static void *
 read_file_to_jni_array_impl(const char *filename)
 {
-    struct dummy_array *array = NULL;
     char *buffer = NULL;
     size_t size = 0;
 
     if (apk_read_file(global.apklib_handle, filename, &buffer, &size) == APK_OK) {
-        array = malloc(sizeof(*array));
-        if (array == NULL)
-            return NULL;
-        array->data = buffer;
-        array->length = size;
-        array->element_size = 1;
+        if(!global.use_dvm)
```

Comment:
This introduces lots of code duplication for handing the non-DVM and DVM case. Are there any situations where we can't use the DVM (or in other words, can we use dvm unconditionally)? In any case, it might make sense to only keep the "proper" JNI variants (`NewByteArray`, `GetByteArrayElements`, `ReleaseByteArrayElements`) and just implement a fallback solution instead of working with the dummy_array here.